python/kvikio/nvcomp_codec.py

```diff
@@ -176,34 +179,36 @@ def decode_batch(
             bufs = [cp.asarray(ensure_contiguous_ndarray_like(b)) for b in bufs]
 
         # Prepare compressed chunks buffers.
-        comp_chunks = np.array([b.data.ptr for b in bufs], dtype=np.uintp)
-        comp_chunk_sizes = np.array([b.size for b in bufs], dtype=np.uint64)
+        comp_chunks = cp.array([b.data.ptr for b in bufs], dtype=cp.uintp)
+        comp_chunk_sizes = cp.array([b.size for b in bufs], dtype=cp.uint64)
 
         # Get uncompressed chunk sizes.
         uncomp_chunk_sizes = self._algo.get_decompress_size(
             comp_chunks,
             comp_chunk_sizes,
             self._stream,
         )
-        # Copy to host since we'll need it to properly allocate buffers.
-        uncomp_chunk_sizes_h = uncomp_chunk_sizes.get()
 
-        max_chunk_size = uncomp_chunk_sizes_h.max()
+        # Check whether the uncompressed chunks are all the same size.
+        # cupy.unique returns sorted sizes.
+        sorted_chunk_sizes = cp.unique(uncomp_chunk_sizes)
+        max_chunk_size = sorted_chunk_sizes[-1].item()
+        is_equal_chunks = sorted_chunk_sizes.shape[0] == 1
 
         # Get temp buffer size.
         temp_size = self._algo.get_decompress_temp_size(num_chunks, max_chunk_size)
 
         temp_buf = cp.empty(temp_size, dtype=cp.uint8)
 
         # Prepare uncompressed chunks buffers.
-        # First, allocate chunks of appropriate sizes and then
+        # First, allocate chunks of max_chunk_size and then
         # copy the pointers to a pointer array in GPU memory as required by nvCOMP.
-        # TODO(akamenev): probably can allocate single contiguous buffer.
-        uncomp_chunks = [
-            cp.empty(size, dtype=cp.uint8) for size in uncomp_chunk_sizes_h
-        ]
-        uncomp_chunk_ptrs = cp.array(
-            [c.data.ptr for c in uncomp_chunks], dtype=cp.uintp
+        # For performance reasons, we use max_chunk_size so we can create
+        # a rectangular array with the same pointer increments.
+        uncomp_chunks = cp.empty((num_chunks, max_chunk_size), dtype=cp.uint8)
+        p_start = uncomp_chunks.data.ptr
+        uncomp_chunk_ptrs = p_start + (
+            cp.arange(0, num_chunks * max_chunk_size, max_chunk_size)
```

Comment:
<!-- Nitpicks are trivial preference-based requests. These should be non-blocking by nature.  -->
**nitpick (non-blocking):** ‏Maybe explicitly set the dtype here:
```suggestion
        uncomp_chunk_ptrs = cupy.uint64(p_start) + (
            cp.arange(0, num_chunks * max_chunk_size, max_chunk_size, dtype=cupy.uint64)
```
Rationale: I am always worried about mixed-kind casting (there are some circumstances, I think not here because `p_start` is just a python integer), but notice that `cupy.arange(10, dtype="uint64") + cupy.arange(10, dtype="int64")` has dtype `float64`!

Reply:
Great catch - fixed! The code worked because I got lucky :) with `p_start` being an integer, as you mentioned.

Reply:
Also, in case of adding a scalar, the resulting type remains `uint64`:
```
assert (p_start + cp.arange(0, num_chunks * max_chunk_size, max_chunk_size, dtype=cp.uint64)).dtype == cp.uint64
```